sandman/source/shell/input_window_eventful_buffer.h

```diff
@@ -0,0 +1,194 @@
+
+/// Fixed sized eventful buffer.
+template <typename CharT, std::size_t kCapacity>
+class Shell::InputWindow::EventfulBuffer
+{
+	static_assert(std::is_integral_v<CharT>);
+
+	public:
+		// Type of internal data buffer.
+		using Data = std::array<CharT, kCapacity>;
+
+		// Can certainly use `CharT` as an alias for `typename Data::value_type`.
+		static_assert(std::is_same_v<CharT, typename Data::value_type>);
+
+		using OnStringUpdateListener = void (*)(typename Data::size_type const index, CharT const character);
+		using OnClearListener = void (*)();
+		using OnDecrementStringLengthListener = void (*)(typename Data::size_type const newStringLength);
+
+	private:
+		// Internal data buffer.
+		Data m_Data{};
+		typename Data::size_type m_StringLength{0u};
+		Common::NonNull<OnStringUpdateListener> m_OnStringUpdate;
+		Common::NonNull<OnClearListener> m_OnClear;
+		Common::NonNull<OnDecrementStringLengthListener> m_OnDecrementStringLength;
+
+	public:
+		static_assert(std::tuple_size_v<Data> > 0u, "Assert can subtract from size without underflow.");
+
+		// The last position is reserved for the null character.
+		static constexpr typename Data::size_type kMaxStringLength{ std::tuple_size_v<Data> - 1u };
+
+		[[gnu::always_inline]] constexpr Data const& GetData() const
+		{
+			return m_Data;
+		}
+
+		[[gnu::always_inline]] constexpr typename Data::size_type GetLength() const
+		{
+			return m_StringLength;
+		}
+
+		explicit constexpr EventfulBuffer() = default;
+
+		// Construct a string with events.
+		// Pass a null pointer to ignore an event.
+		explicit constexpr EventfulBuffer(
+			OnStringUpdateListener          const onStringUpdateListener          ,
+			OnClearListener                 const onClearListener                 ,
+			OnDecrementStringLengthListener const onDecrementStringLengthListener):
+			m_OnStringUpdate                     (onStringUpdateListener         ),
+			m_OnClear                            (onClearListener                ),
+			m_OnDecrementStringLength            (onDecrementStringLengthListener)
+		{}
+
+		// Insert a character at any valid index in the string.
+		// Can insert a character at the end by inserting at
+		// index string length; that is what `EventfulBuffer::Push` does.
+		//
+		// Inserting a character pushes all characters after it one position to the right.
+		constexpr bool Insert(typename Data::size_type const index, CharT const character)
+		{
+			// Can insert at any index in the string or at the end if index is string length.
+			// Can only insert a character if the string is not at maximum capacity.
+			if (index <= m_StringLength and m_StringLength < kMaxStringLength)
+			{
+				// Starting from the index of the null terminator which is at
+				// index string length, will iterate leftward shifting each character
+				// to the right by one position until reached index to insert
+				// the the new character at.
+				for (typename Data::size_type i{ m_StringLength }; i > index; --i)
```

Comment:
It is called for every character assignment because the view of the input string in the shell is updated per individual character rather than rewriting the whole string on every update.